File: apps/clients/python/subsystems/system_info.py
import json
import os
import platform
import subprocess
from pathlib import Path
from typing import Any, Dict

import psutil

from device import (
    Action,
    ActionConfig,
    ActionPayload,
    Monitor,
    MonitorConfig,
    SubsystemDefinition,
    DELETE_FILE_ACTION_ENDPOINT,
    EXECUTE_ACTION_ENDPOINT,
    LIST_DIR_ACTION_ENDPOINT,
    MEMORY_TOTAL_ENDPOINT,
    MEMORY_USAGE_ENDPOINT,
    MKDIR_ACTION_ENDPOINT,
    PRINT_ACTION_ENDPOINT,
    READ_FILE_ACTION_ENDPOINT,
    SYSTEM_SUBSYSTEM_NAME,
    SYSTEM_SUBSYSTEM_TYPE,
    WRITE_FILE_ACTION_ENDPOINT,
)
from mqtt_client import ActionEnvelope
from .utils import (
    entry_type,
    extract_path_from_payload,
    reply_with_data,
    reply_with_error,
    resolve_relative_path,
    target_relative,
)


class SystemInfoTemplate:
    def __init__(self, base_dir: Path) -> None:
        self.base_dir = base_dir

    def build(self, _device_name: str) -> SubsystemDefinition:
        return SubsystemDefinition(
            name=SYSTEM_SUBSYSTEM_NAME,
            type=SYSTEM_SUBSYSTEM_TYPE,
            monitors=[
                MonitorConfig(
                    monitor=Monitor(
                        name="memory_total",
                        label="Total memory",
                        description="Total physical memory detected when the agent booted",
                        units="bytes",
                        endpoint=MEMORY_TOTAL_ENDPOINT,
                        connection_type="mqtt",
                        card_props={"icon": "database", "color": "$green10"},
                        ephemeral=False,
                    ),
                    boot=self.publish_total_memory,
                ),
                MonitorConfig(
                    monitor=Monitor(
                        name="memory_used",
                        label="Used memory",
                        description="Periodically reported RAM usage",
                        units="bytes",
                        endpoint=MEMORY_USAGE_ENDPOINT,
                        connection_type="mqtt",
                        card_props={"icon": "activity", "color": "$blue10"},
                        ephemeral=True,
                    ),
                    interval_seconds=5,
                    tick=self.publish_used_memory,
                ),
                MonitorConfig(
                    monitor=Monitor(
                        name="cpu_model",
                        label="CPU model",
                        description="CPU name/model",
                        endpoint="/system/monitors/cpu_model",
                        connection_type="mqtt",
                        card_props={"icon": "cpu", "color": "$orange10"},
                        ephemeral=False,
                    ),
                    boot=self.publish_cpu_model,
                ),
                MonitorConfig(
                    monitor=Monitor(
                        name="cpu_cores",
                        label="CPU cores",
                        description="Number of logical CPU cores",
                        endpoint="/system/monitors/cpu_cores",
                        connection_type="mqtt",
                        card_props={"icon": "grid", "color": "$purple10"},
                        ephemeral=False,
                    ),
                    boot=self.publish_cpu_cores,
                ),
                MonitorConfig(
                    monitor=Monitor(
                        name="cpu_frequency",
                        label="CPU frequency",
                        description="Current CPU frequency (MHz)",
                        endpoint="/system/monitors/cpu_frequency",
                        connection_type="mqtt",
                        card_props={"icon": "activity", "color": "$pink10"},
                        ephemeral=False,
                    ),
                    boot=self.publish_cpu_frequency,
                ),
                MonitorConfig(
                    monitor=Monitor(
                        name="os_version",
                        label="Operating system",
                        description="Host OS and version",
                        endpoint="/system/monitors/os_version",
                        connection_type="mqtt",
                        card_props={"icon": "monitor", "color": "$cyan10"},
                        ephemeral=False,
                    ),
                    boot=self.publish_os_version,
                ),
            ],
            actions=[
                ActionConfig(
                    action=Action(
                        name="print",
                        label="Print to stdout",
                        description="Send a message that the local agent prints to stdout",
                        endpoint=PRINT_ACTION_ENDPOINT,
                        connection_type="mqtt",
                        payload=ActionPayload(type="string"),
                        card_props={"icon": "terminal"},
                    ),
                    handler=self.handle_print,
                ),
                ActionConfig(
                    action=Action(
                        name="execute",
                        label="Execute command",
                        description="Run a shell command on the host and return its output",
                        endpoint=EXECUTE_ACTION_ENDPOINT,
                        connection_type="mqtt",
                        payload=ActionPayload(type="string"),
                        card_props={"icon": "code", "color": "$red10"},
                        mode="request-reply",
                    ),
                    handler=self.handle_execute,
                ),
                ActionConfig(
                    action=Action(
                        name="list_dir",
                        label="List directory",
                        description="List files in a directory relative to the agent",
                        endpoint=LIST_DIR_ACTION_ENDPOINT,
                        connection_type="mqtt",
                        payload=ActionPayload(
                            type="json-schema",
                            schema={
                                "path": {
                                    "type": "string",
                                    "title": "Directory",
                                    "default": ".",
                                    "description": "Relative path to list",
                                }
                            },
                        ),
                        card_props={"icon": "folder", "color": "$blue9"},
                        mode="request-reply",
                    ),
                    handler=self.handle_list_dir,
                ),
                ActionConfig(
                    action=Action(
                        name="read_file",
                        label="Read file",
                        description="Read a file relative to the agent",
                        endpoint=READ_FILE_ACTION_ENDPOINT,
                        connection_type="mqtt",
                        payload=ActionPayload(
                            type="json-schema",
                            schema={
                                "path": {
                                    "type": "string",
                                    "title": "Path",
                                    "description": "Relative file path",
                                }
                            },
                        ),
                        card_props={"icon": "file-text", "color": "$green9"},
                        mode="request-reply",
                    ),
                    handler=self.handle_read_file,
                ),
                ActionConfig(
                    action=Action(
                        name="write_file",
                        label="Write file",
                        description="Write contents to a file relative to the agent",
                        endpoint=WRITE_FILE_ACTION_ENDPOINT,
                        connection_type="mqtt",
                        payload=ActionPayload(
                            type="json-schema",
                            schema={
                                "path": {
                                    "type": "string",
                                    "title": "Path",
                                    "description": "Relative file path",
                                },
                                "content": {
                                    "type": "string",
                                    "title": "Content",
                                    "description": "Text to write",
                                },
                            },
                        ),
                        card_props={"icon": "edit", "color": "$yellow9"},
                        mode="request-reply",
                    ),
                    handler=self.handle_write_file,
                ),
                ActionConfig(
                    action=Action(
                        name="delete_file",
                        label="Delete file",
                        description="Delete a file relative to the agent",
                        endpoint=DELETE_FILE_ACTION_ENDPOINT,
                        connection_type="mqtt",
                        payload=ActionPayload(
                            type="json-schema",
                            schema={
                                "path": {
                                    "type": "string",
                                    "title": "Path",
                                    "description": "Relative file path",
                                }
                            },
                        ),
                        card_props={"icon": "trash", "color": "$red9"},
                        mode="request-reply",
                    ),
                    handler=self.handle_delete_file,
                ),
                ActionConfig(
                    action=Action(
                        name="mkdir",
                        label="Create directory",
                        description="Create a directory relative to the agent",
                        endpoint=MKDIR_ACTION_ENDPOINT,
                        connection_type="mqtt",
                        payload=ActionPayload(
                            type="json-schema",
                            schema={
                                "path": {
                                    "type": "string",
                                    "title": "Directory",
                                    "description": "Relative directory path",
                                }
                            },
                        ),
                        card_props={"icon": "folder-plus", "color": "$purple9"},
                        mode="request-reply",
                    ),
                    handler=self.handle_mkdir,
                ),
            ],
        )

    def publish_total_memory(self, _ctx, mqtt):
        total = psutil.virtual_memory().total
        mqtt.publish(MEMORY_TOTAL_ENDPOINT, str(total))

    def publish_used_memory(self, _ctx, mqtt):
        used = psutil.virtual_memory().used
        mqtt.publish(MEMORY_USAGE_ENDPOINT, str(used))

    def publish_cpu_model(self, _ctx, mqtt):
        model = _cpu_model()
        mqtt.publish("/system/monitors/cpu_model", model)

    def publish_cpu_cores(self, _ctx, mqtt):
        mqtt.publish("/system/monitors/cpu_cores", str(psutil.cpu_count(logical=True) or 0))

    def publish_cpu_frequency(self, _ctx, mqtt):
        freq = psutil.cpu_freq()
        mhz = freq.current if freq else 0
        mqtt.publish("/system/monitors/cpu_frequency", f"{mhz:.2f}")

    def publish_os_version(self, _ctx, mqtt):
        mqtt.publish("/system/monitors/os_version", f"{platform.system()} {platform.release()}")

    def handle_print(self, msg: ActionEnvelope) -> None:
        raw = (msg.payload or b"").decode("utf-8", errors="ignore").strip()
        if not raw:
            print("[action:print] <empty>")
            return
        try:
            parsed = json.loads(raw)
            print("[action:print] " + json.dumps(parsed, indent=2))
        except Exception:
            print("[action:print] " + raw)

    def handle_execute(self, msg: ActionEnvelope) -> None:
        command = _extract_command((msg.payload or b"").decode("utf-8", errors="ignore"))
        if not command:
            reply_with_error(msg, ValueError("empty command"))
            return
        try:
            completed = subprocess.run(
                command,
                shell=True,
                capture_output=True,
                text=True,
                timeout=120,
            )
            response: Dict[str, Any] = {
                "command": command,
                "output": (completed.stdout or "") + (completed.stderr or ""),
                "exitCode": completed.returncode,
            }
            if completed.returncode != 0:
                response["error"] = "process exited with code {}".format(completed.returncode)
            reply_with_data(msg, response)
        except subprocess.TimeoutExpired:
            reply_with_error(msg, TimeoutError("command timed out"))
        except Exception as err:
            reply_with_error(msg, err)

    def handle_list_dir(self, msg: ActionEnvelope) -> None:
        try:
            path_arg = extract_path_from_payload(msg.payload, ["path", "directory"])
            target = resolve_relative_path(str(self.base_dir), path_arg)
            with os.scandir(target) as entries:
                items = [{"name": e.name, "type": entry_type(e)} for e in entries]
            reply_with_data(msg, {"path": path_arg, "items": items})
        except Exception as err:
            reply_with_error(msg, err)

    def handle_read_file(self, msg: ActionEnvelope) -> None:
        try:
            path_arg = extract_path_from_payload(msg.payload, ["path"])
            target = resolve_relative_path(str(self.base_dir), path_arg)
            content = Path(target).read_text()
            reply_with_data(msg, {"path": path_arg, "content": content})
        except Exception as err:
            reply_with_error(msg, err)

    def handle_write_file(self, msg: ActionEnvelope) -> None:
        try:
            data = json.loads((msg.payload or b"{}"))
            path_arg = data.get("path")
            content = data.get("content", "")
            if not path_arg or not isinstance(path_arg, str):
                raise ValueError("payload must be a JSON object with path and content")
            target = resolve_relative_path(str(self.base_dir), path_arg)
            Path(target).parent.mkdir(parents=True, exist_ok=True)
            Path(target).write_text(content)
            reply_with_data(msg, {"path": target_relative(str(self.base_dir), target), "size": len(content)})
        except Exception as err:
            reply_with_error(msg, err)

    def handle_delete_file(self, msg: ActionEnvelope) -> None:
        try:
            path_arg = extract_path_from_payload(msg.payload, ["path"])
            target = resolve_relative_path(str(self.base_dir), path_arg)
            os.remove(target)
            reply_with_data(msg, {"path": path_arg, "deleted": True})
        except Exception as err:
            reply_with_error(msg, err)

    def handle_mkdir(self, msg: ActionEnvelope) -> None:
        try:
            path_arg = extract_path_from_payload(msg.payload, ["path", "directory"])
            target = resolve_relative_path(str(self.base_dir), path_arg)
            Path(target).mkdir(parents=True, exist_ok=True)
            reply_with_data(msg, {"path": path_arg, "created": True})
        except Exception as err:
            reply_with_error(msg, err)


def _extract_command(payload: str) -> str:
    payload = payload.strip()
    if not payload:
        return ""
    if payload.startswith("{"):
        try:
            data = json.loads(payload)
            cmd = data.get("command")
            if isinstance(cmd, str):
                return cmd.strip()
        except Exception:
            return payload
    return payload


def _cpu_model() -> str:
    try:
        with open("/proc/cpuinfo", "r", encoding="utf-8") as f:
            for line in f:
                if line.lower().startswith("model name"):
                    return line.split(":", 1)[1].strip()
    except Exception:
        pass
    return platform.processor() or platform.machine()

File: apps/clients/go/internal/subsystems/system_info.go
package subsystems

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	"github.com/shirou/gopsutil/v4/cpu"
	"github.com/shirou/gopsutil/v4/host"
	vmem "github.com/shirou/gopsutil/v4/mem"

	"ventoagent/internal/cards"
	"ventoagent/internal/vento"
)

type SystemInfoTemplate struct{}

func NewSystemInfoTemplate() Template {
	return &SystemInfoTemplate{}
}

func (t *SystemInfoTemplate) Build(string) Definition {
	baseDir := getWorkingDir()
	return Definition{
		Name: "system",
		Type: "virtual",
		Monitors: []MonitorConfig{
			// === SISTEMA ===
			{
				Monitor: vento.Monitor{
					Name:           "os_version",
					Label:          "Operating system",
					Description:    "Host OS and version",
					Endpoint:       "/system/monitors/os_version",
					ConnectionType: "mqtt",
					Ephemeral:      false,
					CardProps: map[string]any{
						"icon":  "monitor",
						"color": "$blue10",
						"html":  cards.Text,
						"order": 1,
					},
				},
				Boot: t.publishOSVersion,
			},
			// === CPU ===
			{
				Monitor: vento.Monitor{
					Name:           "cpu_model",
					Label:          "CPU model",
					Description:    "CPU name/model",
					Endpoint:       "/system/monitors/cpu_model",
					ConnectionType: "mqtt",
					Ephemeral:      false,
					CardProps: map[string]any{
						"icon":  "cpu",
						"color": "$orange10",
						"html":  cards.Text,
						"order": 2,
					},
				},
				Boot: t.publishCPUModel,
			},
			{
				Monitor: vento.Monitor{
					Name:           "cpu_cores",
					Label:          "CPU cores",
					Description:    "Number of logical CPU cores",
					Endpoint:       "/system/monitors/cpu_cores",
					ConnectionType: "mqtt",
					Ephemeral:      false,
					CardProps: map[string]any{
						"icon":  "layout-grid",
						"color": "$orange9",
						"order": 3,
					},
				},
				Boot: t.publishCPUCores,
			},
			{
				Monitor: vento.Monitor{
					Name:           "cpu_frequency",
					Label:          "CPU frequency",
					Description:    "Current CPU frequency (MHz)",
					Endpoint:       "/system/monitors/cpu_frequency",
					ConnectionType: "mqtt",
					Ephemeral:      false,
					CardProps: map[string]any{
						"icon":  "activity",
						"color": "$red10",
						"html":  cards.Frequency,
						"order": 4,
					},
				},
				Boot: t.publishCPUFrequency,
			},
			// === MEMORIA ===
			{
				Monitor: vento.Monitor{
					Name:           "memory_total",
					Label:          "Total memory",
					Description:    "Total physical memory detected when the agent booted",
					Units:          "bytes",
					Endpoint:       vento.MemoryTotalEndpoint,
					ConnectionType: "mqtt",
					Ephemeral:      false,
					CardProps: map[string]any{
						"icon":  "database",
						"color": "$green10",
						"html":  cards.Bytes,
						"order": 5,
					},
				},
				Boot: t.publishTotalMemory,
			},
			{
				Monitor: vento.Monitor{
					Name:           "memory_used",
					Label:          "Used memory",
					Description:    "Periodically reported RAM usage",
					Units:          "bytes",
					Endpoint:       vento.MemoryUsageEndpoint,
					ConnectionType: "mqtt",
					Ephemeral:      true,
					CardProps: map[string]any{
						"icon":  "activity",
						"color": "$green9",
						"html":  cards.Bytes,
						"order": 6,
					},
				},
				Interval: 5 * time.Second,
				Tick:     t.publishUsedMemory,
			},
		},
		Actions: []ActionConfig{
			// === DIRECTORIOS (azul) ===
			{
				Action: vento.Action{
					Name:           "list_dir",
					Label:          "List directory",
					Description:    "List files in a directory relative to the agent",
					Endpoint:       vento.ListDirActionEndpoint,
					ConnectionType: "mqtt",
					Payload: vento.ActionPayload{
						Type: "json-schema",
						Schema: map[string]any{
							"path": map[string]any{
								"type":        "string",
								"title":       "Directory",
								"default":     ".",
								"description": "Relative path to list",
							},
						},
					},
					CardProps: map[string]any{
						"icon":  "folder",
						"color": "$blue10",
						"order": 10,
					},
					Mode: "request-reply",
				},
				Handler: func(m vento.ActionEnvelope) error {
					return handleListDirAction(m, baseDir)
				},
			},
			{
				Action: vento.Action{
					Name:           "mkdir",
					Label:          "Create directory",
					Description:    "Create a directory relative to the agent",
					Endpoint:       vento.MkdirActionEndpoint,
					ConnectionType: "mqtt",
					Payload: vento.ActionPayload{
						Type: "json-schema",
						Schema: map[string]any{
							"path": map[string]any{
								"type":        "string",
								"title":       "Directory",
								"description": "Relative directory path",
							},
						},
					},
					CardProps: map[string]any{
						"icon":  "folder-plus",
						"color": "$blue9",
						"order": 11,
					},
					Mode: "request-reply",
				},
				Handler: func(m vento.ActionEnvelope) error {
					return handleMkdirAction(m, baseDir)
				},
			},
			// === ARCHIVOS (verde/amarillo/rojo) ===
			{
				Action: vento.Action{
					Name:           "read_file",
					Label:          "Read file",
					Description:    "Read a file relative to the agent",
					Endpoint:       vento.ReadFileActionEndpoint,
					ConnectionType: "mqtt",
					Payload: vento.ActionPayload{
						Type: "json-schema",
						Schema: map[string]any{
							"path": map[string]any{
								"type":        "string",
								"title":       "Path",
								"description": "Relative file path",
							},
						},
					},
					CardProps: map[string]any{
						"icon":  "file-text",
						"color": "$green10",
						"order": 12,
					},
					Mode: "request-reply",
				},
				Handler: func(m vento.ActionEnvelope) error {
					return handleReadFileAction(m, baseDir)
				},
			},
			{
				Action: vento.Action{
					Name:           "write_file",
					Label:          "Write file",
					Description:    "Write contents to a file relative to the agent",
					Endpoint:       vento.WriteFileActionEndpoint,
					ConnectionType: "mqtt",
					Payload: vento.ActionPayload{
						Type: "json-schema",
						Schema: map[string]any{
							"path": map[string]any{
								"type":        "string",
								"title":       "Path",
								"description": "Relative file path",
							},
							"content": map[string]any{
								"type":        "string",
								"title":       "Content",
								"description": "Text to write",
							},
						},
					},
					CardProps: map[string]any{
						"icon":  "pencil",
						"color": "$yellow10",
						"order": 13,
					},
					Mode: "request-reply",
				},
				Handler: func(m vento.ActionEnvelope) error {
					return handleWriteFileAction(m, baseDir)
				},
			},
			{
				Action: vento.Action{
					Name:           "delete_file",
					Label:          "Delete file",
					Description:    "Delete a file relative to the agent",
					Endpoint:       vento.DeleteFileActionEndpoint,
					ConnectionType: "mqtt",
					Payload: vento.ActionPayload{
						Type: "json-schema",
						Schema: map[string]any{
							"path": map[string]any{
								"type":        "string",
								"title":       "Path",
								"description": "Relative file path",
							},
						},
					},
					CardProps: map[string]any{
						"icon":  "trash",
						"color": "$red10",
						"order": 14,
					},
					Mode: "request-reply",
				},
				Handler: func(m vento.ActionEnvelope) error {
					return handleDeleteFileAction(m, baseDir)
				},
			},
			// === SISTEMA (naranja/púrpura) ===
			{
				Action: vento.Action{
					Name:           "execute",
					Label:          "Execute command",
					Description:    "Run a shell command on the host and return its output",
					Endpoint:       vento.ExecuteActionEndpoint,
					ConnectionType: "mqtt",
					Payload: vento.ActionPayload{
						Type: "string",
					},
					CardProps: map[string]any{
						"icon":  "terminal",
						"color": "$orange10",
						"order": 15,
					},
					Mode: "request-reply",
				},
				Handler: handleExecuteAction,
			},
			{
				Action: vento.Action{
					Name:           "print",
					Label:          "Print to stdout",
					Description:    "Send a message that the local agent prints to stdout",
					Endpoint:       vento.PrintActionEndpoint,
					ConnectionType: "mqtt",
					Payload: vento.ActionPayload{
						Type: "string",
					},
					CardProps: map[string]any{
						"icon":  "message-square",
						"color": "$purple10",
						"order": 16,
					},
				},
				Handler: handlePrintAction,
			},
		},
	}
}

func (t *SystemInfoTemplate) publishCPUModel(ctx context.Context, mqtt *vento.MQTTClient) error {
	info, err := cpu.InfoWithContext(ctx)
	if err != nil || len(info) == 0 {
		return err
	}
	return mqtt.Publish("/system/monitors/cpu_model", info[0].ModelName)
}

func (t *SystemInfoTemplate) publishCPUCores(ctx context.Context, mqtt *vento.MQTTClient) error {
	count, err := cpu.CountsWithContext(ctx, true)
	if err != nil {
		return err
	}
	return mqtt.Publish("/system/monitors/cpu_cores", fmt.Sprintf("%d", count))
}

func (t *SystemInfoTemplate) publishCPUFrequency(ctx context.Context, mqtt *vento.MQTTClient) error {
	info, err := cpu.InfoWithContext(ctx)
	if err != nil || len(info) == 0 {
		return err
	}
	return mqtt.Publish("/system/monitors/cpu_frequency", fmt.Sprintf("%.2f", info[0].Mhz))
}

func (t *SystemInfoTemplate) publishOSVersion(ctx context.Context, mqtt *vento.MQTTClient) error {
	info, err := host.InfoWithContext(ctx)
	if err != nil {
		return err
	}
	return mqtt.Publish("/system/monitors/os_version", fmt.Sprintf("%s %s", info.Platform, info.PlatformVersion))
}

func handlePrintAction(msg vento.ActionEnvelope) error {
	payload := msg.Payload
	trimmed := strings.TrimSpace(string(payload))
	if trimmed == "" {
		fmt.Println("[action:print] <empty>")
		return nil
	}
	var asJSON any
	if json.Unmarshal(payload, &asJSON) == nil {
		if formatted, err := json.MarshalIndent(asJSON, "", "  "); err == nil {
			fmt.Printf("[action:print] %s\n", formatted)
			return nil
		}
	}
	fmt.Printf("[action:print] %s\n", trimmed)
	return nil
}

func (t *SystemInfoTemplate) publishTotalMemory(ctx context.Context, mqtt *vento.MQTTClient) error {
	stats, err := vmem.VirtualMemoryWithContext(ctx)
	if err != nil {
		return err
	}
	return mqtt.Publish(vento.MemoryTotalEndpoint, fmt.Sprintf("%d", stats.Total))
}

func (t *SystemInfoTemplate) publishUsedMemory(ctx context.Context, mqtt *vento.MQTTClient) error {
	stats, err := vmem.VirtualMemoryWithContext(ctx)
	if err != nil {
		return err
	}
	return mqtt.Publish(vento.MemoryUsageEndpoint, fmt.Sprintf("%d", stats.Used))
}

func handleExecuteAction(msg vento.ActionEnvelope) error {
	command := extractCommand(string(msg.Payload))
	if command == "" {
		if msg.CanReply() {
			return msg.ReplyJSON(map[string]any{
				"error": "empty command",
			})
		}
		return errors.New("empty command")
	}

	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
	defer cancel()

	shell, args := commandArgs(command)
	cmd := exec.CommandContext(ctx, shell, args...)
	output, err := cmd.CombinedOutput()

	response := map[string]any{
		"command": command,
		"output":  string(output),
	}

	exitCode := 0
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else if ctx.Err() != nil {
			exitCode = -1
		} else {
			exitCode = 1
		}
		response["error"] = err.Error()
	}
	if ctx.Err() != nil {
		response["timeout"] = true
		if ctx.Err() != context.Canceled {
			response["error"] = ctx.Err().Error()
		}
	}
	response["exitCode"] = exitCode

	if msg.CanReply() {
		if err := msg.ReplyJSON(response); err != nil {
			return err
		}
	}
	return err
}

func commandArgs(command string) (string, []string) {
	if runtime.GOOS == "windows" {
		return "cmd", []string{"/C", command}
	}
	return "sh", []string{"-c", command}
}

func extractCommand(payload string) string {
	trimmed := strings.TrimSpace(payload)
	if trimmed == "" {
		return ""
	}
	if strings.HasPrefix(trimmed, "{") {
		var data map[string]any
		if err := json.Unmarshal([]byte(trimmed), &data); err == nil {
			if cmd, ok := data["command"].(string); ok {
				return strings.TrimSpace(cmd)
			}
		}
	}
	return trimmed
}

func handleListDirAction(msg vento.ActionEnvelope, base string) error {
	pathArg, err := extractPathFromPayload(msg.Payload, "path", "directory")
	if err != nil {
		return replyWithError(msg, err)
	}

	target, err := resolveRelativePath(base, pathArg)
	if err != nil {
		return replyWithError(msg, err)
	}

	entries, err := os.ReadDir(target)
	if err != nil {
		return replyWithError(msg, err)
	}
	items := make([]map[string]any, 0, len(entries))
	for _, entry := range entries {
		item := map[string]any{
			"name": entry.Name(),
			"type": entryType(entry),
		}
		items = append(items, item)
	}
	return replyWithData(msg, map[string]any{
		"path":  pathArg,
		"items": items,
	})
}

func handleReadFileAction(msg vento.ActionEnvelope, base string) error {
	pathArg, err := extractPathFromPayload(msg.Payload, "path")
	if err != nil {
		return replyWithError(msg, err)
	}
	target, err := resolveRelativePath(base, pathArg)
	if err != nil {
		return replyWithError(msg, err)
	}
	data, err := os.ReadFile(target)
	if err != nil {
		return replyWithError(msg, err)
	}
	return replyWithData(msg, map[string]any{
		"path":    pathArg,
		"content": string(data),
	})
}

func handleWriteFileAction(msg vento.ActionEnvelope, base string) error {
	var payload struct {
		Path    string `json:"path"`
		Content string `json:"content"`
	}
	if err := json.Unmarshal(msg.Payload, &payload); err != nil || strings.TrimSpace(payload.Path) == "" {
		return replyWithError(msg, errors.New("payload must be a JSON object with path and content"))
	}
	target, err := resolveRelativePath(base, payload.Path)
	if err != nil {
		return replyWithError(msg, err)
	}
	if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
		return replyWithError(msg, err)
	}
	if err := os.WriteFile(target, []byte(payload.Content), 0o644); err != nil {
		return replyWithError(msg, err)
	}
	return replyWithData(msg, map[string]any{
		"path": targetRelative(base, target),
		"size": len(payload.Content),
	})
}

func handleDeleteFileAction(msg vento.ActionEnvelope, base string) error {
	pathArg, err := extractPathFromPayload(msg.Payload, "path")
	if err != nil {
		return replyWithError(msg, err)
	}
	target, err := resolveRelativePath(base, pathArg)
	if err != nil {
		return replyWithError(msg, err)
	}
	if err := os.Remove(target); err != nil {
		return replyWithError(msg, err)
	}
	return replyWithData(msg, map[string]any{
		"path":    pathArg,
		"deleted": true,
	})
}

func handleMkdirAction(msg vento.ActionEnvelope, base string) error {
	pathArg, err := extractPathFromPayload(msg.Payload, "path", "directory")
	if err != nil {
		return replyWithError(msg, err)
	}
	target, err := resolveRelativePath(base, pathArg)
	if err != nil {
		return replyWithError(msg, err)
	}
	if err := os.MkdirAll(target, 0o755); err != nil {
		return replyWithError(msg, err)
	}
	return replyWithData(msg, map[string]any{
		"path":    pathArg,
		"created": true,
	})
}

func extractPathFromPayload(payload []byte, keys ...string) (string, error) {
	trimmed := strings.TrimSpace(string(payload))
	if trimmed == "" {
		return ".", nil
	}
	if trimmed[0] != '{' {
		return trimmed, nil
	}
	var data map[string]any
	if err := json.Unmarshal(payload, &data); err != nil {
		return "", err
	}
	for _, key := range keys {
		if value, ok := data[key]; ok {
			if str, ok := value.(string); ok {
				return str, nil
			}
		}
	}
	return "", fmt.Errorf("payload missing required path field (%v)", keys)
}

func resolveRelativePath(base, rel string) (string, error) {
	if strings.TrimSpace(rel) == "" {
		rel = "."
	}
	if filepath.IsAbs(rel) {
		return filepath.Clean(rel), nil
	}
	cleaned := filepath.Clean(rel)
	return filepath.Join(base, cleaned), nil
}

func targetRelative(base, full string) string {
	rel, err := filepath.Rel(base, full)
	if err != nil {
		return full
	}
	return rel
}

func entryType(entry fs.DirEntry) string {
	if entry.IsDir() {
		return "directory"
	}
	return "file"
}

func replyWithError(msg vento.ActionEnvelope, err error) error {
	if msg.CanReply() {
		_ = msg.ReplyJSON(map[string]any{
			"error": err.Error(),
		})
	}
	return err
}

func replyWithData(msg vento.ActionEnvelope, data map[string]any) error {
	if msg.CanReply() {
		return msg.ReplyJSON(data)
	}
	return nil
}

func getWorkingDir() string {
	if dir, err := os.Getwd(); err == nil {
		return dir
	}
	return "."
}
